Adding bash completion and helper scripts

leedm777 opened the pull request:
This patch adds some bash helper scripts.

 * docker-machine.bash - command completion for docker-machine
 * docker-machine-prompt.bash - function for putting the active machine
   name in PS1
 * docker-machine-wrapper.bash - function wrapper adding an `eval`
   command that runs `eval $(docker-machine env whatever)` in the
   current shell.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "completion" git@github.com:leedm777/machine.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
Thank you!

/cc @albers

chadxz commented:
+1 for `docker-machine eval <env>`!

sruffell commented:
I'll add my +1 as well....

thaJeztah commented:
I wonder if the helper scripts (prompt, eval) should be in the "completions" directory or moved somewhere else inside "contrib"; I can imagine some people wanting the completions installed, but not those helpers

leedm777 commented:
@thaJeztah I debated that. The prompt function is harmless if you don't put it in your `PS1`. The eval wrapper can be disabled by setting `DOCKER_MACHINE_WRAPPED` to false prior to sourcing the script. That feels "good enough" to me, but I would not object strongly to reorganizing.

The advantage of keeping them together is that if someone wants completion, they _probably_ want the other shell stuff as well. Similarly, if they disable completion then the _probably_ don't want the other shell bloat.

FWIW, Git [put its prompt script in its completion directory](https://github.com/git/git/tree/master/contrib/completion), which made me feel better about doing it this way.

dmp42 commented:
@nathanleclaire wdyt?

dmp42 commented:
Also, #1309 was attempted before. Anything to merge from there?

thaJeztah commented:
@leedm777 thanks, I'm not a maintainer in this repository, and no strong opinion, apart from trying to keep consistency between the various docker repositories. Just thought I'd mention it.

Thanks so much though, for creating this!

nathanleclaire commented:
I'm interested (and thanks for putting this together @leedm777 !), need a chance to review a bit more thoroughly, but two things stand out as needing improvement:

1. I'd prefer `docker-machine use` to `docker-machine eval`
2. It needs documentation, however, I'm not sure the right place (Installation?) so I'm open to suggestions there.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "completion" git@github.com:leedm777/machine.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

leedm777 commented:
@nathanleclaire I've added docs, and changed `eval` to `use`. Any further thoughts?

nathanleclaire commented on docs/install-machine.md:
```diff
@@ -57,6 +57,21 @@ to your PATH.
 
+## Installing bash completion scripts
+
+There are a [number of scripts](../contrib/completion/bash) which add command
```
I'm not sure if this link is correct - @moxiegirl can you comment?  It's meant to link to source code in the docker machine repo.

nathanleclaire replied:
Likewise, would be good to get your input on this doc in general.

leedm777 replied:
@nathanleclaire The link work when viewed [on my fork](https://github.com/leedm777/machine/blob/completion/docs/install-machine.md). I'll gladly take recommendations if there's a better way to link to the scripts.

moxiegirl replied:
@leedm777 When you run `make docs` to validate your changes --- you are doing this right? :-D --- it reports an error on your link.

```
...
ERROR: 2015/10/22 LinkResolver: No page found for "../contrib/completion/bash" on page "machine/install-machine.md".
...
```
Soon our own @SvenDowideit will add this to the CI and such things will be caught when you submit your PR. For now, you have to run `make docs`

@nathanleclaire I'm checking the added content. Thanks for the ping.

nathanleclaire commented on contrib/completion/bash/docker-machine-prompt.bash:
```diff
@@ -0,0 +1,46 @@
+#  1a. Copy this file somewhere and source it in your .bashrc
+#      source /some/where/docker-machine-prompt.bash
+#  1b. Alternatively, just copy this file into into /etc/bash_completion.d
+#  2. Change your PS1 to call __git_ps1 as command-substitution
```
Looks like this still references `git`?  Copy paste error I guess?

nathanleclaire commented:
Putting `docker-machine status` in PS1 is going to make the terminal pretty slow, so I'm hesitant to recommend it directly.

albers commented:
@thaJeztah Thanks for pointing me to this, and sorry for the delay.
I'm not familiar with docker-machine because I could not get it working in several previous attempts. I'll go for it next weekend.
@leedm777 Thanks for your work. On first sight it looks very impressive!

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "completion" git@github.com:leedm777/machine.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
@albers thanks for looking; I am happy already with you just giving it a quick look to see if there are possible issues or inconsistencies with the other projects :+1:

moxiegirl commented on docs/install-machine.md:
```diff
@@ -57,6 +57,21 @@ to your PATH.
         $ docker-machine -v
         machine version 0.5.0 (3e06852)
 
+## Installing bash completion scripts
```
@leedm777 Nicely done.  I would do a few things to sort this. Put your link OUT last   --- no sense inducing them to switch channels before you've explained yourself.  Break out the functionality into a scannable list.

---> https://gist.github.com/moxiegirl/9b2dba37dc8e75f5654e

## Installing bash completion scripts

The machine repo supplies several scripts] that add features such as:

* command completion
* a function that displays the active machine in your shell prompt
* a function wrapper that adds a `docker-machine use` subcommand

The `docker-machine use function executes `eval "$(docker-machine env X)"` in the current shell.

To install the scripts, copy or link them into your `/etc/bash_completion.d` or
`/usr/local/etc/bash_completion.d` file. To enable the `docker-machine` shell prompt, add `$(__docker-machine-ps1)` to your `PS1` setting in `~/.bashrc`.

    PS1='[\u@\h \W$(__docker-machine-ps1)]\$ '

You can find additional documentation in the comments at the [top of each script](https://github.com/docker/machine/tree/master/contrib/scripts/bash).

leedm777 commented:
@moxiegirl Is that better :-)

moxiegirl commented:
@leedm777 thank you. Very glad to help your really helpful contribution!  LGTM

nathanleclaire commented on docs/install-machine.md:
```diff
@@ -57,6 +57,27 @@ to your PATH.
 
+## Installing bash completion scripts
+
+The Machine repository supplies several `bash` scripts that add features such
```
Let's move the link to the `contrib/completion/bash` section of the repository to this line, i.e. "supplies several `bash` scripts" should be the link to there.

nathanleclaire replied:
Ah, I see it's the opposite of what @moxiegirl recommended.  No need to change it then.

nathanleclaire commented on docs/install-machine.md:
```diff
@@ -57,6 +57,27 @@ to your PATH.
+
+* command completion
+* a function that displays the active machine in your shell prompt
+* a function wrapper that adds a `docker-machine use` subcommand
```
I'd extend it to "subcommand to switch the active machine"

nathanleclaire commented on docs/install-machine.md:
```diff
@@ -57,6 +57,27 @@ to your PATH.
+* a function that displays the active machine in your shell prompt
+* a function wrapper that adds a `docker-machine use` subcommand
+
+The `docker-machine` function executes `eval "$(docker-machine env X)"` in the
```
I'd say:

> The `docker-machine use` function added by the completion scripts expands `docker-machine use X` to `eval "$(docker-machine env X)`.

leedm777 replied:
@nathanleclaire With your recommended change above, I think this line can go away entirely.

nathanleclaire replied:
Hm, I'd prefer it to be explicit, since "switching the active machine" has meant different things at different points in Machine's history.

nathanleclaire commented:
OK looks pretty good, I have a few docs nits, and just one more suggestion for the UX:  If you type `docker-machine use` by itself, it should spit back the help text just like `docker-machine use --help` does.

leedm777 commented:
I think I've addressed all the feedback. This history was getting ugly, so I rebased on master and squashed to a single commit.

Thanks to everyone for all the feedback!

thaJeztah commented:
Great work @leedm777 !

nathanleclaire commented:
LGTM

albers commented:
OK, so there won' be any more need for my feedback.

thaJeztah commented:
Sorry for that @albers, I wasn't aware @nathanleclaire was gonna merge before you had the chance to give feedback :blush:
